File: zig/ui/glfw/vertex_buffer_layout.zig
const std = @import("std");
const ArrayList = std.ArrayList;
const gl = @cImport(@cInclude("GL/glew.h"));


pub const VertexBufferElement = struct {
    count: u32,
    _type: u32,
    normalized: bool,

    pub fn bufferElementGetSize(self: *VertexBufferElement) u32 {
        switch(@as(i32, @intCast(self._type))) {
            gl.GL_FLOAT => {
                return @sizeOf(gl.GLfloat);
            },
            gl.GL_UNSIGNED_INT => {
                return @sizeOf(gl.GLuint);
            },
            gl.GL_INT => {
                return @sizeOf(gl.GLint);
            },
            gl.GL_UNSIGNED_BYTE => {
                return @sizeOf(gl.GLbyte);
            },
            else => {}
        }
        return 0;
    }
};



pub const VertexBufferLayout = struct {
    elements: ArrayList(VertexBufferElement),
    stride: u32,
    allocator: std.mem.Allocator,

    pub fn destroy(self: *VertexBufferLayout) void {
        self.elements.deinit(self.allocator);
    }

    pub fn pushf32(self: *VertexBufferLayout, count: u32, normalized: bool) void {
        self.elements.append(VertexBufferElement{.count = count, ._type = gl.GL_FLOAT, .normalized = normalized});
        self.stride += VertexBufferElement.bufferElementGetSize(gl.GL_FLOAT) * count;
    }

    pub fn pushu32(self: *VertexBufferLayout, count: u32, normalized: bool) void {
        self.elements.append(VertexBufferElement{.count = count, ._type = gl.GL_UNSIGNED_INT, .normalized = normalized});
        self.stride += VertexBufferElement.bufferElementGetSize(gl.GL_UNSIGNED_INT) * count;
    }

    pub fn pushu8(self: *VertexBufferLayout, count: u32, normalized: bool) void {
        self.elements.append(VertexBufferElement{.count = count, ._type = gl.GL_UNSIGNED_BYTE, .normalized = normalized});
        self.stride += VertexBufferElement.bufferElementGetSize(gl.GL_UNSIGNED_BYTE) * count;
    }
};


pub fn vertexBufferLayout(allocator: std.mem.Allocator) !VertexBufferLayout {
    const vbl: VertexBufferLayout = .{.elements = try ArrayList(VertexBufferElement).initCapacity(allocator, 0),
                                   .stride = 0, .allocator = allocator};

    return vbl;
}

File: c++/ui/opengl/vertex_buffer_layout.cpp
#include "vertex_buffer_layout.hpp"

VertexBufferElement::VertexBufferElement(u32 count, u32 _type, bool normalized) : count(count), _type(_type), normalized(normalized) {}
VertexBufferElement::~VertexBufferElement(){}

u32 VertexBufferElement::bufferElementGetSize() {
    switch(this->_type) {
        case GL_FLOAT: {
            return sizeof(GLfloat);
        } break;
        case GL_UNSIGNED_INT: {
            return sizeof(GLuint);
        } break;
        case GL_INT: {
            return sizeof(GLint);
        } break;
        case GL_UNSIGNED_BYTE: {
            return sizeof(GLbyte);
        } break;
        default: return 0;
    }
    return 0;
}


VertexBufferLayout::VertexBufferLayout() {
    stride = 0;
}

VertexBufferLayout::~VertexBufferLayout() {}

void VertexBufferLayout::pushf32(u32 count, bool normalized) {
    VertexBufferElement e{count, GL_FLOAT, normalized};
    this->elements.push_back(e);
    this->stride += e.bufferElementGetSize() * count;
}

void VertexBufferLayout::pushu32(u32 count, bool normalized) {
    VertexBufferElement e{count, GL_UNSIGNED_INT, normalized};
    this->elements.push_back(e);
    this->stride += e.bufferElementGetSize() * count;
}

void VertexBufferLayout::pushu8(u32 count, bool normalized) {
    VertexBufferElement e{count, GL_UNSIGNED_BYTE, normalized};
    this->elements.push_back(e);
    this->stride += e.bufferElementGetSize() * count;
}
